external/esp_idf_port/spi_flash/cache_utils.c

```diff
@@ -0,0 +1,343 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdlib.h>
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+#include <sched.h>
+
+#include <rom/spi_flash.h>
+#include <rom/cache.h>
+#include <chip/esp32_soc.h>
+#include <chip/dport_access.h>
+#include "esp_ipc.h"
+#include "esp_attr.h"
+#include "esp_intr_alloc.h"
+#include "esp_spi_flash.h"
+
+static void IRAM_ATTR spi_flash_disable_cache(uint32_t cpuid, uint32_t *saved_state);
+static void IRAM_ATTR spi_flash_restore_cache(uint32_t cpuid, uint32_t saved_state);
+
+static uint32_t s_flash_op_cache_state[2];
+
+#ifdef CONFIG_SMP
+static SemaphoreHandle_t s_flash_op_mutex;
+static volatile bool s_flash_op_can_start = false;
+static volatile bool s_flash_op_complete = false;
+#ifndef NDEBUG
+static volatile int s_flash_op_cpu = -1;
+#endif
+
+void spi_flash_init_lock(void)
+{
+	s_flash_op_mutex = xSemaphoreCreateRecursiveMutex();
+	assert(s_flash_op_mutex != NULL);
+}
+
+void spi_flash_op_lock(void)
+{
+	xSemaphoreTakeRecursive(s_flash_op_mutex, portMAX_DELAY);
+}
+
+void spi_flash_op_unlock(void)
+{
+	xSemaphoreGiveRecursive(s_flash_op_mutex);
+}
+
+/*
+ If you're going to modify this, keep in mind that while the flash caches of the pro and app
+ cpu are separate, the psram cache is *not*. If one of the CPUs returns from a flash routine
+ with its cache enabled but the other CPUs cache is not enabled yet, you will have problems
+ when accessing psram from the former CPU.
+*/
+
+void IRAM_ATTR spi_flash_op_block_func(void *arg)
+{
+	// Disable scheduler on this CPU
+	sched_lock();
+	// Restore interrupts that aren't located in IRAM
+	esp_intr_noniram_disable();
+	uint32_t cpuid = (uint32_t) arg;
+	// s_flash_op_complete flag is cleared on *this* CPU, otherwise the other
+	// CPU may reset the flag back to false before IPC task has a chance to check it
+	// (if it is preempted by an ISR taking non-trivial amount of time)
+	s_flash_op_complete = false;
+	s_flash_op_can_start = true;
+	while (!s_flash_op_complete) {
+		// busy loop here and wait for the other CPU to finish flash operation
+	}
+	// Flash operation is complete, re-enable cache
+	spi_flash_restore_cache(cpuid, s_flash_op_cache_state[cpuid]);
+	// Restore interrupts that aren't located in IRAM
+	esp_intr_noniram_enable();
+	// Re-enable scheduler
+	sched_unlock();
+}
+
+void IRAM_ATTR spi_flash_disable_interrupts_caches_and_other_cpu(void)
+{
+	spi_flash_op_lock();
+
+	const uint32_t cpuid = xPortGetCoreID();
+	const uint32_t other_cpuid = (cpuid == 0) ? 1 : 0;
+#ifndef NDEBUG
+	// For sanity check later: record the CPU which has started doing flash operation
+	assert(s_flash_op_cpu == -1);
+	s_flash_op_cpu = cpuid;
+#endif
+
+	if (xTaskGetSchedulerState() == taskSCHEDULER_NOT_STARTED) {
```

Comment:
many freertos APIs in this file, so it is not compiled?
if not, we can remove it ?

Reply:
no need to remove, this code is prepared for SMP( more cpus)